akka-actor-typed-tests/src/test/scala/docs/akka/typed/RouterSpec.scala

```diff
@@ -143,5 +143,51 @@ class RouterSpec extends ScalaTestWithActorTestKit("akka.loglevel=warning") with
 
       probe.receiveMessages(10)
     }
+    "show group routing with consistent hashing" in {
+
+      val probe1 = createTestProbe[String]()
+      val probe2 = createTestProbe[String]()
+
+      object Proxy {
+
+        val RegisteringKey = ServiceKey[Command]("aggregator-key")
+
+        def mapping(command: Command) = command.id
+
+        sealed trait Command {
+          def id: String
+        }
+        case class Message(id: String, content: String) extends Command
+
+        def apply(monitor: ActorRef[String]): Behavior[Command] =
+          Behaviors.setup[Command] { context =>
+            context.system.receptionist ! Receptionist.Register(RegisteringKey, context.self)
+            Behaviors.receiveMessage {
+              case Message(id, _) =>
+                monitor ! id
+                Behaviors.same
+            }
+          }
+      }
+      //they register to Proxy.RegisteringKey on initialization
+      spawn(Proxy(probe1.ref))
+      spawn(Proxy(probe2.ref))
+
+      //messages sent to a router with constant hashing
+      val router = spawn(Routers.group(Proxy.RegisteringKey).withConsistentHashingRouting(10, Proxy.mapping))
+      router ! Proxy.Message("123", "Text1")
```

Comment:
The group router defers routing until it has seen the first listing but it is possible that that first listing is before the proxies has registered, so this test is not deterministic - it could happen that these four messages all end up in deadletters on a bad day.

Because of that it would be better to not register inside the proxy but do it in the test, and wait for registration completion (`Register(key, actor, replyTo)`) before spawning the router, that way we know that the first listing to the group will contain both proxy actors. (BTW the tests above has that same issue since they don't wait for receptionist response)

Reply:
Thanks @johanandren , very interesting insight. I'll implement your suggestion.

Reply:
@johanandren please, let me know if you agree with the approach.

Reply:
👍